File: security-assessment-dashboard/frontend/src/types/target.ts
export type TargetType = "ipv4" | "ipv6" | "cidr" | "hostname" | "domain" | "url";

export interface Target {
  id: string;
  assessment_id: string;
  target_type: TargetType;
  target_value: string;
  resolved_ip: string | null;
  notes: string | null;
  enabled: boolean;
  created_at: string;
  updated_at: string;
}

export interface TargetCreatePayload {
  target_type: TargetType;
  target_value: string;
  notes?: string | null;
}

export interface TargetUpdatePayload {
  target_type?: TargetType;
  target_value?: string;
  notes?: string | null;
  enabled?: boolean;
}

export interface TargetValidateResult {
  valid: boolean;
  normalized_value: string | null;
  message: string | null;
}

export interface TargetImportError {
  line_number: number;
  raw_value: string;
  reason: string;
}

export interface TargetBulkImportResult {
  total_lines: number;
  imported: number;
  skipped_duplicates: number;
  skipped_invalid: number;
  errors: TargetImportError[];
  imported_targets: Target[];
}

export const TARGET_TYPE_OPTIONS: { value: TargetType; label: string }[] = [
  { value: "ipv4", label: "IPv4" },
  { value: "ipv6", label: "IPv6" },
  { value: "cidr", label: "CIDR" },
  { value: "hostname", label: "Hostname" },
  { value: "domain", label: "Domain" },
  { value: "url", label: "URL" },
];

export interface TargetListParams {
  search?: string;
  target_type?: TargetType;
  enabled?: boolean;
  sort_by?: string;
  sort_dir?: "asc" | "desc";
  page?: number;
  page_size?: number;
}

File: security-assessment-dashboard/backend/schemas/target.py
"""Request/response schemas for targets."""

import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from backend.models.enums import TargetType


class TargetCreate(BaseModel):
    """Payload to add a target to an assessment."""

    target_type: TargetType
    target_value: str = Field(min_length=1, max_length=512)
    notes: str | None = None


class TargetUpdate(BaseModel):
    """Payload to partially update a target. Omitted fields are left unchanged."""

    target_type: TargetType | None = None
    target_value: str | None = Field(default=None, min_length=1, max_length=512)
    notes: str | None = None
    enabled: bool | None = None


class TargetRead(BaseModel):
    """A target as returned by the API."""

    model_config = {"from_attributes": True}

    id: uuid.UUID
    assessment_id: uuid.UUID
    target_type: TargetType
    target_value: str
    resolved_ip: str | None
    notes: str | None
    enabled: bool
    created_at: datetime
    updated_at: datetime


class TargetDuplicateRequest(BaseModel):
    """Optional explicit value for a duplicated target.

    IPv4/IPv6/CIDR targets have no safe auto-generated variant (the service
    refuses to fabricate a nearby-but-different address), so ``target_value``
    is required for those types and optional for hostname/domain/URL, where
    the service can auto-suffix a value to keep it unique.
    """

    target_value: str | None = Field(default=None, max_length=512)


class TargetValidateRequest(BaseModel):
    """A target value to check without saving it."""

    target_type: TargetType
    target_value: str = Field(min_length=1, max_length=512)


class TargetValidateResponse(BaseModel):
    """Result of a dry-run target validation."""

    valid: bool
    normalized_value: str | None = None
    message: str | None = None


class TargetImportError(BaseModel):
    """One rejected line from a bulk import."""

    line_number: int
    raw_value: str
    reason: str


class TargetBulkImportResult(BaseModel):
    """Summary of a bulk target import."""

    total_lines: int
    imported: int
    skipped_duplicates: int
    skipped_invalid: int
    errors: list[TargetImportError]
    imported_targets: list[TargetRead]
